plugins/chain_plugin/trx_finality_status_processing.cpp

```diff
@@ -1,12 +1,215 @@
+         }
+         else {
+            _storage.insert(finality_status_object{.trx_id = trx_id, .trx_expiry = std::get<1>(trx_tuple)->expiration(), .received = now, .block_id = block_id});
+         }
+      }
+      if (modified) {
+         const auto& indx = _storage.index().get<by_block_num>();
+         auto iter = indx.cbegin();
+         if (iter != indx.cend()) {
+            _last_tracked_block_id = iter->block_id;
+         }
+      }
+
+      if (bsp) {
+         _last_proc_block_num = chain::block_header::num_from_id(_head_block_id);
+      }
+   }
+
+   void trx_finality_status_processing_impl::handle_rollback() {
+      const auto& indx = _storage.index().get<by_block_num>();
+      std::deque<chain::transaction_id_type> trxs;
+      auto iter = indx.lower_bound(_head_block_id);
+      std::transform(iter, indx.end(), trxs.end(), []( const finality_status_object& obj ) { return obj.trx_id; });
+      for (const auto& trx_id : trxs) {
+         auto trx_iter = _storage.find(trx_id);
+         _storage.modify( trx_iter, []( finality_status_object& obj ) {
+            obj.block_id = chain::block_id_type{};
+            obj.forked_out = true;
+         } );
+      }
+   }
+
+   bool trx_finality_status_processing_impl::status_expiry_of_trxs(const fc::time_point& now) {
+#warning TODO change by_success_status_expiry to by_status_expiry and remove by_fail_status_expiry
+      const auto& indx = _storage.index().get<by_success_status_expiry>();
+      std::deque<chain::transaction_id_type> remove_trxs;
+      const fc::time_point success_expiry = now - _success_duration;
+
+      // just need to have any valid block id, since ordering is based off of empty block id or non-empty
+      const auto& a_valid_block_id = _head_block_id;
+
+      // find the successful (in any block) transactions that are past the successful expiry time
+      auto iter = indx.lower_bound(boost::make_tuple(a_valid_block_id, success_expiry));
+      auto get_id = []( const finality_status_object& obj ) { return obj.trx_id; };
+      std::transform(iter, indx.end(), remove_trxs.end(), get_id);
+
+      // find the failure (not in a block) transactions that are past the failure expiry time
+      const fc::time_point failure_expiry = now - _failure_duration;
+      // need to verify not in a block since the index will transition from old failed transactions to newer successful transactions
+      for (iter = indx.lower_bound(boost::make_tuple(chain::block_id_type{}, failure_expiry));
+           iter != indx.end() && iter->block_id == chain::block_id_type{};
+           ++iter) {
+         remove_trxs.push_back(iter->trx_id);
+      }
+
+      for (const auto& trx_id : remove_trxs) {
+         _storage.erase(trx_id);
+      }
+      return !remove_trxs.empty();
+   }
+
+   void trx_finality_status_processing_impl::ensure_storage(const fc::time_point& now) {
+#warning TODO logging
+      const int64_t remaining_storage = _storage.memory_size() - _max_storage;
```

Comment:
Shouldn't this be `_max_storage - _storage.memory_size()`

Reply:
yes